File: server/routes/bolsa.mjs
// ─────────────────────────────────────────────────────────────────────
// A Bolsa: cotações ao vivo, através do servidor.
//
// O browser não fala com a fonte directamente — a política de segurança
// não deixa, e é assim que deve ser. Pede aqui, e o servidor vai buscar
// ao Yahoo Finance (o gráfico intradiário, que não pede chave nenhuma)
// e guarda a resposta um minuto: cem visitantes a olhar para o mesmo
// título são um pedido lá fora, não cem.
//
//   GET /api/bolsa?s=AAPL,MSFT&r=1d     as cotações e a série de cada um
//
// Nada do que vem de fora entra sem ser conferido: os símbolos têm uma
// forma fixa, o intervalo é de uma lista, e o que se devolve ao browser
// é só o que a aplicação usa — números e nomes, não o JSON todo.
// ─────────────────────────────────────────────────────────────────────
import { LIMITS, STOCKS } from '../config.mjs';
import { json } from '../http.mjs';
import { bump, ipKey } from '../security.mjs';

const SYMBOL_RE = /^[A-Z0-9][A-Z0-9.=^-]{0,11}$/;

/** Os intervalos que a aplicação mostra, e como se pedem. */
const RANGES = {
  '1d': { range: '1d', interval: '5m' },
  '1w': { range: '5d', interval: '30m' },
  '1m': { range: '1mo', interval: '1d' },
  '3m': { range: '3mo', interval: '1d' },
  '1y': { range: '1y', interval: '1wk' },
};

const cache = new Map();

/** Só o que interessa, do que a fonte devolve. */
function shape(symbol, body) {
  const result = body && body.chart && body.chart.result && body.chart.result[0];
  if (!result || !result.meta) return null;
  const meta = result.meta;
  const closes = ((result.indicators && result.indicators.quote && result.indicators.quote[0]) || {}).close || [];
  const stamps = result.timestamp || [];
  const points = [];
  for (let i = 0; i < closes.length; i++) {
    if (typeof closes[i] === 'number') points.push([stamps[i] || 0, Number(closes[i].toFixed(4))]);
  }
  const price = typeof meta.regularMarketPrice === 'number' ? meta.regularMarketPrice : points.length ? points[points.length - 1][1] : null;
  const previous =
    typeof meta.chartPreviousClose === 'number'
      ? meta.chartPreviousClose
      : typeof meta.previousClose === 'number'
        ? meta.previousClose
        : null;
  if (price == null) return null;
  return {
    symbol,
    name: String(meta.shortName || meta.longName || symbol).slice(0, 60),
    currency: String(meta.currency || '').slice(0, 8),
    price,
    previous,
    change: previous != null ? Number((price - previous).toFixed(4)) : null,
    percent: previous ? Number((((price - previous) / previous) * 100).toFixed(2)) : null,
    market: String(meta.marketState || '').slice(0, 12),
    points,
  };
}

async function fetchOne(symbol, range) {
  const key = symbol + ':' + range;
  const hit = cache.get(key);
  if (hit && Date.now() - hit.at < STOCKS.ttl) return hit.data;
  const r = RANGES[range];
  const url =
    STOCKS.source + encodeURIComponent(symbol) + '?range=' + r.range + '&interval=' + r.interval + '&includePrePost=false';
  try {
    const res = await fetch(url, {
      headers: { 'User-Agent': STOCKS.agent, Accept: 'application/json' },
      signal: AbortSignal.timeout(STOCKS.timeout),
    });
    if (!res.ok) throw new Error('http ' + res.status);
    const data = shape(symbol, await res.json());
    if (!data) throw new Error('forma');
    cache.set(key, { at: Date.now(), data });
    if (cache.size > STOCKS.cacheMax) cache.delete(cache.keys().next().value);
    return data;
  } catch (_) {
    // Uma resposta velha vale mais do que nenhuma.
    return hit ? hit.data : null;
  }
}

export async function handleBolsa(req, res, url) {
  const key = ipKey(req);
  if (!bump('bolsa:' + key, LIMITS.stocksPerIpWindow, LIMITS.stocksPerIp))
    return json(res, 429, { ok: false, error: 'limite' });
  if (!bump('bolsa:global', LIMITS.stocksGlobalWindow, LIMITS.stocksGlobal))
    return json(res, 429, { ok: false, error: 'limite' });

  const range = RANGES[url.searchParams.get('r') || '1d'] ? url.searchParams.get('r') || '1d' : '1d';
  const symbols = String(url.searchParams.get('s') || '')
    .toUpperCase()
    .split(',')
    .map((s) => s.trim())
    .filter((s) => SYMBOL_RE.test(s))
    .slice(0, LIMITS.stocksPerRequest);
  if (!symbols.length) return json(res, 400, { ok: false, error: 'simbolos' });

  const quotes = (await Promise.all(symbols.map((s) => fetchOne(s, range)))).filter(Boolean);
  return json(res, 200, { ok: true, range, at: Date.now(), quotes });
}

File: api/app/routers/bolsa.py
# ─────────────────────────────────────────────────────────────────────
# A Bolsa: cotações ao vivo.
#
# Ao contrário do antigo servidor Node, que falava com esta API por HTTP
# numa rede interna, agora corre no mesmo processo — o `yfinance` é
# bloqueante, por isso cada chamada vai para um thread à parte
# (`asyncio.to_thread`) para não travar o resto da API enquanto espera
# pelo Yahoo.
#
#   GET /api/bolsa?s=AAPL,MSFT&r=1d   as cotações e a série de cada um
#   GET /api/bolsa/detalhe?s=AAPL     estatísticas, perfil e notícias
#   GET /api/bolsa/procurar?q=apple   símbolos por nome ou sigla
#
# Nada do que vem de fora entra sem ser conferido: os símbolos têm uma
# forma fixa, o intervalo é de uma lista, e o que se devolve ao browser
# é só o que a aplicação usa.
# ─────────────────────────────────────────────────────────────────────
import asyncio
import re
import time

from fastapi import APIRouter, Request
from starlette.responses import JSONResponse

from app.bolsa.client import RANGES, chart_for, details_for, quotes_for, search_for
from app.config import LIMITS
from app.security import bump, ip_key
from app.validation import one_line

router = APIRouter()

# O primeiro carácter também pode ser '^': é assim que o Yahoo escreve um
# índice — '^GSPC' é o S&P 500, e está na lista por omissão.
SYMBOL_RE = re.compile(r"^[A-Z0-9^][A-Z0-9.=^-]{0,11}$")


def _range_of(request: Request) -> str:
    r = request.query_params.get("r") or "1d"
    return r if r in RANGES else "1d"


def _symbols_of(raw: str) -> list[str]:
    seen: list[str] = []
    for part in raw.upper().split(","):
        s = part.strip()
        if s and SYMBOL_RE.match(s) and s not in seen:
            seen.append(s)
    return seen[: LIMITS.stocks_per_request]


def _within_limits(request: Request, name: str, window: float, limit: int) -> bool:
    key = ip_key(request)
    if not bump(f"{name}:{key}", window, limit):
        return False
    return bump(f"{name}:global", LIMITS.stocks_global_window, LIMITS.stocks_global)


@router.get("/api/bolsa")
async def bolsa(request: Request) -> JSONResponse:
    if not _within_limits(request, "bolsa", LIMITS.stocks_per_ip_window, LIMITS.stocks_per_ip):
        return JSONResponse({"ok": False, "error": "limite"}, status_code=429)

    range_ = _range_of(request)
    symbols = _symbols_of(request.query_params.get("s", ""))
    if not symbols:
        return JSONResponse({"ok": False, "error": "simbolos"}, status_code=400)

    if range_ == "1d":
        quotes = await asyncio.to_thread(quotes_for, symbols)
    else:
        results = await asyncio.gather(*(asyncio.to_thread(chart_for, s, range_) for s in symbols))
        quotes = [q for q in results if q]

    return JSONResponse({"ok": True, "range": range_, "at": int(time.time() * 1000), "source": "yfinance", "quotes": quotes})


@router.get("/api/bolsa/detalhe")
async def bolsa_detalhe(request: Request) -> JSONResponse:
    if not _within_limits(request, "bolsa", LIMITS.stocks_per_ip_window, LIMITS.stocks_per_ip):
        return JSONResponse({"ok": False, "error": "limite"}, status_code=429)
    symbols = _symbols_of(request.query_params.get("s", ""))
    if not symbols:
        return JSONResponse({"ok": False, "error": "simbolos"}, status_code=400)
    data = await asyncio.to_thread(details_for, symbols[0])
    if not data:
        return JSONResponse({"ok": False, "error": "simbolo"}, status_code=404)
    return JSONResponse({"ok": True, **data})


@router.get("/api/bolsa/procurar")
async def bolsa_procurar(request: Request) -> JSONResponse:
    if not _within_limits(request, "bolsa-procura", LIMITS.stocks_search_window, LIMITS.stocks_search_per_ip):
        return JSONResponse({"ok": False, "error": "limite"}, status_code=429)
    q = one_line(request.query_params.get("q"), LIMITS.stocks_query)
    if not q:
        return JSONResponse({"ok": False, "error": "procura"}, status_code=400)
    results = await asyncio.to_thread(search_for, q)
    return JSONResponse({"ok": True, "results": results})
